### DISCUSSION TITLE
Is creating providers programmatically supported?

### DISCUSSION THREAD

**Discussion by plammens** [rrousselGit/riverpod#1680] [Q&A] [answered]
While reading the introductory parts of the documentation, I've only seen providers being created as global variables:

```dart
// top-level
final myProvider = Provider(...);

class MyClass {
  ...
}
```

Is creating providers programmatically supported?

-----------

For example, suppose we're building a to-do app. This app creates `Task` objects, which are mutable objects that encapsulate a to-do:

```dart
class Task {
  Task({required this.title});

  String title;
  bool completed = false;
}
```

The user can edit the title of a task and mark a task as completed. In order to rebuild the UI when such a change occurs, we make `Task` extend `ChangeNotifier`:

```dart
class Task extends ChangeNotifier {
  Task({required this.title});

  String _title;
  String get title => _title;
  set title(String value) {
    _title = value;
    notifyListeners();
  }

  bool _completed = false;
  bool get completed => _completed;
  set completed(bool value) {
    if (value != _completed) {
      _completed = value;
      notifyListeners();
    }
  }
}
```

One of the screens of the app is a list showing all the saved tasks. Each task is shown in a `Card` widget of its own. 

```dart
class TaskCard extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Card(...);
  }
}
```

When a task is updated (e.g. its title is edited), only the corresponding card should be rebuilt, it would be a waste to rebuild the entire list.

How would you do this in riverpod? My intuition tells me that I could do the following. 

Every time a task is created, we create a provider only for that specific task object.

```dart
// somewhere inside an "add task" button callback
var task = Task(...);
var provider = ChangeNotifierProvider(create: (ref) => task);  
```

Then the specific provider for a specific task could be passed to the corresponding `TaskCard` as a parameter:

```dart
class TaskCard extends StatelessWidget {
  TaskCard({required this.taskProvider});

  final ChangeNotifierProvider<Task> taskProvider;

  @override
  Widget build(BuildContext context) {
    final task = taskProvider.watch();
    return Card(...);
  }
}
```

Then my question is: is this supported? Can I create a provider programmatically and pass it only to the widget instances that need it? 
(Some further questions if this is the case: in the example above, if the task object is disposed, can I safely dispose of the provider?)

-----

If this is not supported, I cannot figure out the "proper" way of implementing the above scenario. Maybe with `.family`? But with what would I parametrise the family, with the task object itself? Probably not because task objects are mutable, and the documentation for `.family` discourages the use of mutable parameters for provider families.

**Comment by TimWhiting** [comment]
Typically you would use a family with the task's id as the parameter.

**Comment by rrousselGit** [comment, marked as answer]
> Is creating providers programmatically supported?

No it isn't.

Providers must be globals. But you can use things like family or scoping to do advanced behaviors, while still defining the provider globally.

**Reply by TimWhiting** [reply]
I just realized that I forgot to finish the new pages of documentation around scoping. I'll try to open some PRs over the weekend or next week to improve documentation in this area.
